sensing/autoware_pointcloud_preprocessor/src/outlier_filter/polar_voxel_outlier_filter_node.cpp

```diff
@@ -0,0 +1,1087 @@
+  stat.add("Warning Threshold", filter_ratio_warn_threshold_);
+  stat.add("Filtering Mode", use_return_type_classification_ ? "Advanced" : "Simple");
+  stat.add("Visibility Only", visibility_estimation_only_ ? "Yes" : "No");
+}
+
+void PolarVoxelOutlierFilterComponent::validate_filter_inputs(
+  const PointCloud2ConstPtr & input, const IndicesPtr & indices)
+{
+  validate_indices(indices);
+  validate_input_cloud(input);
+  validate_required_fields(input);
+}
+
+void PolarVoxelOutlierFilterComponent::validate_indices(const IndicesPtr & indices)
+{
+  if (indices) {
+    RCLCPP_WARN(get_logger(), "Indices are not supported and will be ignored");
+  }
+}
+
+void PolarVoxelOutlierFilterComponent::validate_input_cloud(const PointCloud2ConstPtr & input)
+{
+  if (!input) {
+    RCLCPP_ERROR(get_logger(), "Input point cloud is null");
+    throw std::invalid_argument("Input point cloud is null");
+  }
+}
+
+void PolarVoxelOutlierFilterComponent::validate_required_fields(const PointCloud2ConstPtr & input)
+{
+  validate_return_type_field(input);
+  validate_intensity_field(input);
+}
+
+void PolarVoxelOutlierFilterComponent::validate_return_type_field(const PointCloud2ConstPtr & input)
+{
+  if (!use_return_type_classification_) {
+    return;
+  }
+
+  if (!has_field(input, "return_type")) {
+    RCLCPP_ERROR(
+      get_logger(),
+      "Advanced mode (use_return_type_classification=true) requires 'return_type' field. "
+      "Set use_return_type_classification=false for simple mode or ensure input has return_type "
+      "field.");
+    throw std::invalid_argument("Advanced mode requires return_type field");
+  }
+}
+
+void PolarVoxelOutlierFilterComponent::validate_intensity_field(const PointCloud2ConstPtr & input)
+{
+  if (!has_field(input, "intensity")) {
+    RCLCPP_ERROR(get_logger(), "Input point cloud must have 'intensity' field");
+    throw std::invalid_argument("Input point cloud must have intensity field");
+  }
+}
+
+bool PolarVoxelOutlierFilterComponent::has_field(
+  const PointCloud2ConstPtr & input, const std::string & field_name)
+{
+  for (const auto & field : input->fields) {
+    if (field.name == field_name) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void PolarVoxelOutlierFilterComponent::create_output(
+  const PointCloud2ConstPtr & input, const ValidPointsMask & valid_points_mask,
+  PointCloud2 & output)
+{
+  if (visibility_estimation_only_) {
+    create_empty_output(input, output);
+  } else {
+    create_filtered_output(input, valid_points_mask, output);
+  }
+}
+
+void PolarVoxelOutlierFilterComponent::create_empty_output(
+  const PointCloud2ConstPtr & input, PointCloud2 & output)
+{
+  setup_output_header(output, input, 0);
+}
+
+void PolarVoxelOutlierFilterComponent::conditionally_publish_noise_cloud(
+  const PointCloud2ConstPtr & input, const ValidPointsMask & valid_points_mask)
+{
+  if (publish_noise_cloud_) {
+    publish_noise_cloud(input, valid_points_mask);
+  }
+}
+
```

Comment:
I feel like this is a bit too fine-grained. Removing this function and replacing its call sites with direct
```cpp
if (publish_noise_cloud_) { publish_noise_cloud(...); }
```
would be simpler.

Reply:
Fixed in https://github.com/autowarefoundation/autoware_universe/pull/10996/commits/2378c01835c92fea40dd8e3cf580ca0bb366a1c7